### PR TITLE
Feat/patch management

### PR DESCRIPTION
**Pull request by lukasrothenberger**
- adds patch generator
- adds patch applicator
- adds line_mapping initialization & bookkeeping

### CODE REVIEW

**Comment by goerlibe** [comment]
hardcoded `clang` and `clang++` strings in CodeGenerator cause failure for me.
On my side I changed it to `clang++-11` so I can continue on the GUI, but we should read it from our build_config

**Comment by goerlibe** [comment]
it would be helpful if the patch applicator would return an error code != 0 if applying the suggestions was not successful.
For now I will look if stdout contains "not successful."

**Comment by lukasrothenberger** [comment]
> it would be helpful if the patch applicator would return an error code != 0 if applying the suggestions was not successful. For now I will look if stdout contains "not successful."

@goerlibe 
what should be returned in case an already applied suggestion is to be applied? 
0, since no issue arises, or 1 since it "can not" be applied?

**Comment by lukasrothenberger** [comment]
> hardcoded `clang` and `clang++` strings in CodeGenerator cause failure for me. On my side I changed it to `clang++-11` so I can continue on the GUI, but we should read it from our build_config

@goerlibe 
proposed fix in 123370f5e0045331f58406b8b1320d69594ec989

**Comment by goerlibe** [comment]
rollback of simple_gpu resulted in a wrong line_mapping with  the start_line being mapped to -1

**Comment by goerlibe** [comment]
I can currently not apply more than one suggestion at once, not sure if I am doing something wrong, though

**Comment by goerlibe** [comment]
it would be nice if the patch_generator generates the applie_suggestions.json with an empty array.

btw is there a reason for the entries being strings? they could be simple numbers, right?

File: discopop_library/LineMapping/diff_modifications.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    for diff_entry in cleaned_diff:
+        if "a" in diff_entry:
+            lhs = diff_entry[: diff_entry.index("a")]
+            if "," in lhs:
+                base_line = int(lhs.split(",")[0])
+            else:
+                base_line = int(lhs)
+            rhs = diff_entry[diff_entry.index("a") + 1 :]
+            if "," in rhs:
+                added_lines_count = int(rhs.split(",")[1]) - int(rhs.split(",")[0])
```
Comment (lukasrothenberger) [inline discopop_library/LineMapping/diff_modifications.py:78]:
```suggestion
                added_lines_count = int(rhs.split(",")[1]) - int(rhs.split(",")[0]) + 1
```

File: discopop_library/LineMapping/diff_modifications.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            lhs = diff_entry[: diff_entry.index("c")]
+            if "," in lhs:
+                base_line = int(lhs.split(",")[0])
+                deleted_lines_count = int(lhs.split(",")[1]) - int(lhs.split(",")[0])
+            else:
+                base_line = int(lhs)
+                deleted_lines_count = 1
+            rhs = diff_entry[diff_entry.index("c") + 1 :]
+            if "," in rhs:
+                added_lines_count = int(rhs.split(",")[1]) - int(rhs.split(",")[0])
```
Comment (lukasrothenberger) [inline discopop_library/LineMapping/diff_modifications.py:115]:
```suggestion
                added_lines_count = int(rhs.split(",")[1]) - int(rhs.split(",")[0]) + 1
```

**Review by goerlibe — APPROVED** [review]

**Review by goerlibe — APPROVED** [review]

**Pull request merged**